packages/signers/src/__tests__/sign-transaction-test.ts

```diff
@@ -0,0 +1,550 @@
+import 'test-matchers/toBeFrozenObject';
+
+import { Address } from '@solana/addresses';
+import { SignatureBytes } from '@solana/keys';
+import { CompilableTransaction, IFullySignedTransaction, ITransactionWithSignatures } from '@solana/transactions';
+
+import {
+    partiallySignTransactionWithSigners,
+    signAndSendTransactionWithSigners,
+    signTransactionWithSigners,
+} from '../sign-transaction';
+import {
+    createMockTransactionCompositeSigner,
+    createMockTransactionModifyingSigner,
+    createMockTransactionPartialSigner,
+    createMockTransactionSendingSigner,
+    createMockTransactionWithSigners,
+} from './__setup__';
+
+describe('partiallySignTransactionWithSigners', () => {
+    it('signs the transaction with its extracted signers', async () => {
+        expect.assertions(3);
+
+        // Given a transaction with two signers A and B in its account metas.
+        const signerA = createMockTransactionModifyingSigner('1111' as Address);
+        const signerB = createMockTransactionPartialSigner('2222' as Address);
+        const transaction = createMockTransactionWithSigners([signerA, signerB]);
+
+        // And given signer A and B are mocked to provide the following signatures.
+        const modifiedTransaction = { ...transaction, signatures: { '1111': '1111_signature' } };
+        signerA.modifyAndSignTransactions.mockResolvedValueOnce([modifiedTransaction]);
+        signerB.signTransactions.mockResolvedValueOnce([{ '2222': '2222_signature' }]);
+
+        // When we partially sign this transaction.
+        const signedTransaction = await partiallySignTransactionWithSigners(transaction);
+
+        // Then it contains the expected signatures.
+        expect(signedTransaction.signatures).toStrictEqual({
+            '1111': '1111_signature',
+            '2222': '2222_signature',
+        });
+
+        // And the signers were called with the expected parameters.
+        expect(signerA.modifyAndSignTransactions).toHaveBeenCalledWith([transaction]);
+        expect(signerB.signTransactions).toHaveBeenCalledWith([modifiedTransaction]);
+    });
+
+    it('signs the transaction with extra signers', async () => {
```

Comment:
It enables anyone to use the Signer API without necessarily having constructed their transaction in a way that stores signers in the their account metas.

For instance, you're sending your partially signed transaction over the wire to your server. The server decodes the transaction and, of course, doesn't have signers attached to the account metas. Then the server can still use the same Signer API regardless like so.

```ts
await signTransactionWithSigners(decodedTransaction, [myServerSigner]);
```

Additionally, it allows more complex signing scenarios.

For instance, ephemeral signers.

```ts
// Some factory function offered by some walllet.
const myOriginalSigner = createSignerWithEphemeralSignersSupport();

// Let's imagine that the created signer offers an
// additional method to create these ephemeral signers.
const [
  myEphemeralSigner1, 
  myEphemeralSigner2
] = await myOriginalSigner.createEphemeralSigners(2);

const instruction = await someInstruction({
  signerAccount1: myEphemeralSigner1,
  signerAccount2: myEphemeralSigner2,
})

const transaction = pipe(
  createTransaction({ version: 0 }),
  tx => appendTransactionInstruction(tx, instruction),
  // ...
);
```

In this example, the ephemeral signers could be implemented in different ways:

- If the wallet decides to implement the ephemeral signers as `NoopSigner`, then the transaction won't be signed unless we provide the original signer which will provide signatures for all ephemeral signers.
  ```ts
  await signAndSendTransactionWithSigners(transaction, [myOriginalSigner]);
  ```
- If the wallet decides to implement the ephemeral signers such that they simply defer to the implementation of the original signer, then we don't need to provide an extra signer.
  ```ts
  await signAndSendTransactionWithSigners(transaction);
  ```

Reply:
The issue with that solution is it assumes the provided signer is present in the transaction. It sounds counter-intuitive but it may not be. If you see the example above with ephemeral signers, the `originalSigner` is signing on behalf of its ephemeral signers but is not itself part of the transaction (which would defeat the purpose of using ephemeral signers in the first place).

Another advantage of that optional `signers` array is we can make it so they override the extracted signer in case of duplicates.

If you _reaaaaally_ hate it, we can remove it and add it later if/when people need it but IMO it will be needed for some niche use-cases.

I even think @buffalojoec had some use-cases for that because I got some love when I mentioned I was gonna allow people to inject more signers. 😅 

![CleanShot 2023-11-17 at 22 23 12@2x](https://github.com/solana-labs/solana-web3.js/assets/3642397/f55ab5cf-423b-4cae-8998-0cbba56a3fd6)

Reply:
Yeah I felt like it was useful. Specifically I think I was again leaning toward flexibility on transaction requests, but perhaps it's not the only way that can be done.

Reply:
Basically we can either:
- Remove this option of manually provided signers (which I don't recommend) or
- Fail when more than one sending signer is provided.

But not both. Otherwise, we disable ephemeral signers (see https://github.com/solana-labs/solana-web3.js/pull/1792#discussion_r1397980537).

Reply:
> …If you see the example above with ephemeral signers, the originalSigner is signing on behalf of its ephemeral signers but is not itself part of the transaction…

In the case that you didn't add the `Signer` in the first place, wouldn't you just `addSignersToTransaction()` with the ephemeral signer that, presumably, does match some address in the transaction, and delegates to `originalSigner` through 〰️magic〰️?

I think I don't properly understand what ephemeral signers are. I'm operating on the simp assumption that every signer is associated with an address and knows how to sign bytes, the end. I must be missing something.

Reply:
> delegates to originalSigner through 〰️magic〰️

Yes, but that won't work if there's more than one ephemeral signer for the same original signer and if that original signer is a sending signer. _Unless_, we allow multiple sending signers and choose one at random.

Reply:
Resolving this as decisions were made [here](https://github.com/solana-labs/solana-web3.js/pull/1792#issuecomment-1820487421).

Whilst these decisions make it more difficult for the specific use-case "ephemeral sending signer", a workaround is to make all ephemeral signers `NoopSigner` aside from one, which delegates to the original signer.